src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/FileUtilities.cs

```diff
@@ -1,13 +1,52 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System;
 using System.Globalization;
 using System.IO;
 
 namespace Microsoft.AspNetCore.Razor.Utilities;
 
 internal static class FileUtilities
 {
+    private const string RazorExtension = ".razor";
+    private const string CSHtmlExtension = ".cshtml";
+
+    public static bool IsAnyRazorFilePath(ReadOnlySpan<char> filePath, StringComparison comparison)
+    {
+        return IsRazorComponentFilePath(filePath, comparison) ||
+               IsMvcFilePath(filePath, comparison);
+    }
+
+    public static bool IsRazorComponentFilePath(ReadOnlySpan<char> filePath, StringComparison comparison)
+    {
+        return AdjustToUsableFilePath(filePath).EndsWith(RazorExtension, comparison);
+    }
+
+    public static bool IsMvcFilePath(ReadOnlySpan<char> filePath, StringComparison comparison)
+    {
+        return AdjustToUsableFilePath(filePath).EndsWith(CSHtmlExtension, comparison);
+    }
+
+    private static ReadOnlySpan<char> AdjustToUsableFilePath(ReadOnlySpan<char> filePath)
+    {
+        // In VS Code we sometimes get odd uris with query string components as the file path, for example on the left side of
+        // a diff view. In those cases Roslyn will create a document and the file path will be set to the full raw Uri send via
+        // VS Code. When trying to find out the file extension for those Uris, we need to strip off the query string.
+        //
+        // For example we get:
+        //   git:/c:/Users/dawengie/source/repos/razor01/Pages/Index.cshtml?%7B%22path%22:%22c:%5C%5CUsers%5C%5Cdawengie%5C%5Csource%5C%5Crepos%5C%5Crazor01%5C%5CPages%5C%5CIndex.cshtml%22,%22ref%22:%22~%22%7D
+        //
+        // Given colons and question marks are unlikely, or illegal, file path characters the risk of false positives here is hopefully low.
+        if (filePath.IndexOf(":/") > 0 &&
+           filePath.IndexOf('?') is int realPathEnd and > 0)
```

Comment:
Should this check that `:/` occurs before the `?`?  What happens to long file paths when they're "normalized"? Do they look like this: `/?/D:/very long path`?

Reply:
https://github.com/dotnet/razor/pull/12171